addons/stock/models/stock_move_line.py

```diff
@@ -535,14 +535,37 @@ def _get_similar_move_lines(self):
 
     def _create_and_assign_production_lot(self):
         """ Creates and assign new production lots for move lines."""
-        lot_vals = [{
-            'company_id': ml.move_id.company_id.id,
-            'name': ml.lot_name,
-            'product_id': ml.product_id.id,
-        } for ml in self]
+        lot_vals = []
+        grouped_mls_tracked_by_lot = defaultdict(lambda: self.env['stock.move.line'])
+        mls_tracked_by_sn = self.env['stock.move.line']
+        for ml in self:
+            if ml.product_id.tracking == 'lot':
+                lot_name, product_id, company_id = ml.lot_name, ml.product_id.id, ml.company_id.id
+                # Avoids to add a second time the same lot in the list of lots to create.
+                if not grouped_mls_tracked_by_lot.get((lot_name, product_id, company_id), False):
+                    lot_vals.append({
+                        'company_id': ml.move_id.company_id.id,
+                        'name': ml.lot_name,
+                        'product_id': ml.product_id.id,
+                    })
+                grouped_mls_tracked_by_lot[(lot_name, product_id, company_id)] |= ml
+            else:
+                lot_vals.append({
+                    'company_id': ml.move_id.company_id.id,
+                    'name': ml.lot_name,
+                    'product_id': ml.product_id.id,
+                })
+                mls_tracked_by_sn |= ml
+
         lots = self.env['stock.production.lot'].create(lot_vals)
-        for ml, lot in zip(self, lots):
+        for ml in mls_tracked_by_sn:
+            lot = lots.filtered(lambda lot: lot.product_id.id == ml.product_id.id and lot.name == ml.lot_name and lot.company_id.id == ml.company_id.id)
```

Comment:
I am afraid of performance: Imagine you have 10K serial number (it is a case truly), you will search our lot in 10K lots for each move line (> 10K) == 100000000 operation.